config/auth_server.go

```diff
@@ -0,0 +1,161 @@
+	// If zero the token is valid forever.
+	ExpiresOn json.Number `json:"expires_on"`
+}
+
+// Credentials provider that fetches a token from a locally running HTTP server.
+// The credentials provider will perform a GET request to the configured URL.
+//
+// The header "Metadata: true" will be added to the request, which must be
+// verified by server to prevent SSRF attacks
+//
+// The server is expected to return a JSON response with the following fields:
+//
+//	host: Optional URL of the Databricks host to connect to.
+//	access_token: The requested access token.
+//	token_type: The type of token, which is a "Bearer" access token.
+//	expires_on: The timespan when the access token expires.
+type ServerCredentials struct{}
+
+func (c ServerCredentials) Name() string {
+	return "auth-server"
+}
+
+func (c ServerCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.AuthServerUrl == "" {
+		return nil, nil
+	}
+
+	parsedAuthServerUrl, err := url.Parse(cfg.AuthServerUrl)
+	if err != nil {
+		return nil, fmt.Errorf("invalid auth server URL: %w", err)
+	}
+
+	// only allow localhost URLs
+	if parsedAuthServerUrl.Hostname() != "localhost" && parsedAuthServerUrl.Hostname() != "127.0.0.1" {
+		return nil, fmt.Errorf("invalid auth server URL: %s", cfg.AuthServerUrl)
+	}
+
+	resp, err := makeRequest(cfg.AuthServerUrl)
+	if err != nil {
+		return nil, nil
+	}
+
+	if resp == nil {
+		return nil, nil
+	}
+
+	if resp.Host == "" {
+		return nil, nil
+	}
+
+	if cfg.Host != "" && cfg.Host != resp.Host {
+		return nil, nil
+	}
+
+	cfg.Host = resp.Host
+
+	ts := authServerTokenSource(cfg.AuthServerUrl)
+	return refreshableVisitor(&ts), nil
+}
+
+// makes a request to the server and returns the token
+func makeRequest(serverUrl string) (*serverResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), authServerTimeout)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, serverUrl, nil)
+	if err != nil {
+		return nil, fmt.Errorf("token request: %w", err)
+	}
+	req.Header.Add("Metadata", "true")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("token response: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == 404 {
+		return nil, nil
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("token read: %w", err)
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("token error: %s", body)
+	}
+
+	var response serverResponse
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("token parse: %w", err)
+	}
+
+	// parse and normalize url
+	parsedHost, err := url.Parse(response.Host)
+	if err != nil {
+		return nil, fmt.Errorf("token parse: Invalid URL %s", response.Host)
+	}
+
+	response.Host = parsedHost.String()
+
+	return &response, nil
+}
+
+type authServerTokenSource string
```

Comment:
could you refactor `azureMsiTokenSource` and make it customizable to the extent you need it to be customizable? i suspect those to share 80-90% of the code.